### PR TITLE
RFD - Unified Resource view for web UI

### PR DESCRIPTION
**Pull request by avatus**
Updates #27462

Reader view [here](https://github.com/gravitational/teleport/blob/rfd/0137-unified-resource-view/rfd/0137-unified-resource-view.md)

Let me know what details I might have missed in this RFD as I'm a bit in the weeds and might be taking some knowledge for granted

### CODE REVIEW

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:29]: I still have a branch floating around locally that did use it's own grpc call. I had the same idea as you did, but for the opposite argument
"It's doing almost the same exact thing as ListResources but just adding one more "Kind" to the mix". 

the complexity added to `ListResources` is pretty negligible and doesn't change much. At least, it doesn't change anything that wouldn't already change if we added a _different_ kind like... idk, Locks or whatever.

I'll inspect the old way I did it (separate) now that I have context around keeping it integrated and revisit this

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:61]: Another option would be to use the empty interface that `Items` uses already in this `ListResourcesResponse` and then use the correct `MakeServer` `MakeDatabase` for each type. Not sure which I like more. Both seem a bit meh

Comment (avatus) [reply rfd/0137-unified-resource-view.md]: Ok, I ended up removing this. In our `apiserver.clusterGetUnifiedResources`, I've just skipped the `MakeUnifiedResource` step. It was there to emulate how the other resources were made, but I've just opted to switch/case through the returned `page.Resources` and `MakeX` depending on what it was. I will add the `Kind` field to the UI structs returned by these that way we know what they are on the frontend but besides that, it's a lot cleaner now.

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:27]: I will be using a new rpc endpoint instead, `ListUnifiedResources`

File: rfd/0137-unified-resource-view.md
Comment (kimlisa) [inline rfd/0137-unified-resource-view.md:117]: i never worked with infinite scrolling, but will we be off loading the previously fetched resources?

File: rfd/0137-unified-resource-view.md
Comment (kimlisa) [inline rfd/0137-unified-resource-view.md:104]: just curious how rbac plays into this. what if a user didn't have access to for example nodes? how will we let the user know that they don't have access? (do we want to?)

i think previously, if a user didn't have "list" verb, the tab won't render. I think we are trying to move away from "silently" not rendering things and be more loud about it (eg: you don't have access to nodes, reach out to your teleport admin to request additional perm etc) but i'm not sure

File: rfd/0137-unified-resource-view.md
Comment (kimlisa) [inline rfd/0137-unified-resource-view.md:112]: I'm having trouble envisioning drop down filter for labels and addresses as well. Is there a design for that? Would it be made searchable? There's just going to be a ton of them so I'm curious to know how it will be implemented (if it will be).

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:117]: We will not be offloading previously fetch resources. Just appending new ones at the bottom of the list

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:14]: The checkboxes on the card indicate the card is selected to perform some sort of batch activity, although I'm not sure which batch activities are available atm. cc: @roraback

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:104]: Currently, we just show what is available to user. I believe it's a "silent" fail if they don't have access to something like Kubes. cc: @roraback

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:112]: I don't think the designs are final (yet?) but you can see the current work here https://www.figma.com/file/JKd0Bgs30AnNfNCLUuzzc2/Unified-Resources?type=design&node-id=0-1&mode=design&t=Z6fdPQrsVJWx6CrU-0

cc: @roraback

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:38]: `name/type` will allow us to sort as expected. If we did it `type/name` then all apps would show up first (if I'm understanding that correctly).

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:117]: The designs still call for a total count, and although we are using infinite scrolling in the frontend, the grpc call is still a "pagination" request, just always going forward so the total count is still returned. cc: @roraback any thoughts on removing total count?

File: rfd/0137-unified-resource-view.md
Comment (bl-nero) [reply rfd/0137-unified-resource-view.md:38]: The problem is that we need to apply sorting after all items are retrieved from the watcher anyway. So whether we use this or another type of key doesn't really matter. I think that using a BTree here will only pay off after we store items and search indices separately.

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:112]: Update: the dropdowns for labels and addresses will function mostly as an 'input' box that can be used to guide the user toward creating advance search predicate. imagine, an input box on labels for "key" and then it will allow you to type a "value". They are coming after resource pinning so not including in the initial release.

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:38]: Any change to the sort type/direction, or change to the search query/filters/etc will result in a reset of the search. essentially, its back to "page 1", similar to pagination. sorting in this case will inform the order of results that you can then infinitely scroll through. Our infinite scroll is really just pagination behind the scenes, but only "going to next page" and appending the results.

File: rfd/0137-unified-resource-view.md
Comment (avatus) [reply rfd/0137-unified-resource-view.md:104]: This feature is being killed on our current servers page and also will not be present in the Unified Resources page

Comment (roraback) [reply rfd/0137-unified-resource-view.md:132]: Correct. Users could still discover types through the filter options, and if they filter or search for a type, such as kubernetes, the empty state should give them an indication. The current blank resource screen doesn't seem particularly insightful in the current state, anyway: it just tells them that _either_ there is nothing to see _or_ they don't have permissions, which they could probably figure out for themselves if they know anything about their company's infrastructure. Please LMK if I'm missing something, though!

This does remind me that design for the empty state was one of the casualties of time constraints—we should be careful about our messaging to ensure it doesn't give false impressions.

File: rfd/0137-unified-resource-view.md
Comment (bl-nero) [reply rfd/0137-unified-resource-view.md:25]: @avatus Could you look once more to my comment above? Since we ended up sorting in the API layer anyway, it's now just an arbitrary decision that doesn't have anything to do with sorting. I'd remove this justification from here.

File: rfd/0137-unified-resource-view.md
Comment (bl-nero) [inline rfd/0137-unified-resource-view.md:23]: Is it still valid in the world where we have a separate API request? I think we left the `KindUnifiedResource`, but it became an internal implementation detail (one which, arguably, should be cleaned out so that it doesn't confuse future maintainers).

**Review by bl-nero — APPROVED** [review]
Approving, but please consider my comments about `KindUnifiedResource` and the relationship between key structure and sort order.

**Review by kimlisa — APPROVED** [review]
lgtm and inline with what I saw on the frontend stuff, but i'd wait for @rosstimothy's approval

File: rfd/0137-unified-resource-view.md
Comment (zmb3) [inline rfd/0137-unified-resource-view.md:25]: Does the in-memory store live on the auth server or on the proxy?

**Review by zmb3 — APPROVED** [review]
Looks good.

I would add a section for next steps / potential improvements, as I think we do have some performance concerns for very large clusters and leaving some context behind will help whoever ends up working on improving performance.

**Pull request merged**
